tap_kafka/common.py

```diff
@@ -34,6 +34,8 @@ def generate_catalog(kafka_config) -> list:
         pk_config = kafka_config.get('primary_keys', [])
         if isinstance(pk_config, object):
             pks = list(pk_config.keys())
+    else:  # Add message_key as a default if no custom PKs specified
+        pks = ['message_key']
```

Comment:
Here you default to adding `message_key` to the schema but messages don't necessarily have a key, hence the condition `elif message.key(): # In absence of custom PKs, try to use the message one` in the other file

Reply:
> I'm not aware whether empty PK can cause side-effects for target, though.

It will cause issues, singer records have to match their schema, if the target receives a schema where it was mentioned that a `message_key` property is present and is a PK but then starts receiving records that either don't have this field or it is empty/null then the target fails.
We just fixed this same issue with custom pk in https://github.com/transferwise/pipelinewise-tap-kafka/pull/70

> One of the solutions to this problem would be to make message key completely optional via config, e.g. "use_message_key" bool and only in case of true add it to the schema and try to retrieve it from the message, wdyt?

Sounds like a good option, to add to it, if this flag is true but `message.key()` is null then tap must fail.
wdyt @koszti ?

Reply:
Using message key in topics is highly recommended and it's optional only if the topic is not partitioned. If it's partitioned then message key is required. So the scenario when message key is not available is very rare and not recommended anyway.

I'd make PK mandatory and would not complicate it with an extra `use_message_key`, which will not be used anyway:
1. Use the optional `primary_key` if it exists in the config
2. Default to the message key if no custom `primary_key` in the config
3. Raise exception if no message key in the message

Reply:
@koszti 
> it's optional only if the topic is not partitioned
not really. Kafka can handle partitioning without a key, it will just do so in round-robin fashion. Where it's absolutely required is when compaction is used.

I see value in both approaches, it is indeed rather unlikely to encounter a topic where message wouldn't contain message key. However, I can also see value in giving an option, should need be and the target can coop with absence of key and the exactly-once is not a requirement.

We could make the "use_message_key" by default true, for example. More so, we could always expose message key, so that the only thing such configuration property would influence is schema record (and whether exception should be raised on missing message key).

In case in light of this we opt out for the enforced key, then only error-handling would really change in this PR and type of change would still be breaking.

Reply:
addressed by [4a35b10](https://github.com/transferwise/pipelinewise-tap-kafka/pull/68/commits/4a35b1085b17c1e8c103e2594b4049271d10ff2e)